README.md
<h1 align="center">
  CandyGraph
</h1>

<div align="center">
  A flexible and fast-by-default 2D plotting library tuned for rendering huge datasets on the GPU at interactive speeds.
</div>

<div align="center">
  <h3 align="center">
    <a href="https://wwwtyro.github.io/candygraph/tutorial/dist/">Tutorial</a> |
    <a href="https://wwwtyro.github.io/candygraph/examples/dist/">Examples</a> |
    <a href="https://wwwtyro.github.io/candygraph/api/">API Docs</a>
  </h3>
</div>

- Adopts D3's elegant concept of scales, but implements them on the GPU to maximize data throughput and handle scale changes with no additional cost.
- Performs all rendering on the GPU, even axes and text. This keeps rendering and compositing fast and simple - no need to overlay a canvas or svg over your plot.
- Provides a small but growing set of highly optimized primitives that are tuned for handling huge datasets.
- Provides a set of higher-level composite objects composed of primitives to simplify common rendering tasks.

> CandyGraph is still in [major version zero](https://semver.org/#spec-item-4); the API should be considered unstable.

## Example

<p align="center">
  <img src="media/simple-plot.png">
</p>

```typescript
import CandyGraph, {
  createDefaultFont,
  LinearScale,
  OpaqueLineStrip,
  OrthoAxis,
  CartesianCoordinateSystem,
} from "candygraph";

async function main() {
  const cg = new CandyGraph();
  cg.canvas.width = cg.canvas.height = 384;

  // Generate some x & y data.
  const xs = [];
  const ys = [];
  for (let x = 0; x <= 1; x += 0.001) {
    xs.push(x);
    ys.push(0.5 + 0.25 * Math.sin(x * 2 * Math.PI));
  }

  // Create a viewport. Units are in pixels.
  const viewport = {
    x: 0,
    y: 0,
    width: cg.canvas.width,
    height: cg.canvas.height,
  };

  // Create a coordinate system from two linear scales. Note
  // that we add 32 pixels of padding to the left and bottom
  // of the viewport, and 16 pixels to the top and right.
  const coords = new CartesianCoordinateSystem(
    cg,
    new LinearScale([0, 1], [32, viewport.width - 16]),
    new LinearScale([0, 1], [32, viewport.height - 16])
  );

  // Load the default Lato font
  const font = await createDefaultFont(cg);

  // Clear the viewport.
  cg.clear([1, 1, 1, 1]);

  // Render the a line strip representing the x & y data, and axes.
  cg.render(coords, viewport, [
    new OpaqueLineStrip(cg, xs, ys, {
      colors: [1, 0.5, 0.0],
      widths: 3,
    }),
    new OrthoAxis(cg, coords, "x", font, {
      labelSide: 1,
      tickOffset: -2.5,
      tickLength: 6,
      tickStep: 0.2,
      labelFormatter: (n) => n.toFixed(1),
    }),
    new OrthoAxis(cg, coords, "y", font, {
      tickOffset: 2.5,
      tickLength: 6,
      tickStep: 0.2,
      labelFormatter: (n) => n.toFixed(1),
    }),
  ]);

  // Copy the plot to a new canvas and add it to the document.
  document.body.appendChild(cg.copyTo(viewport));
}

main();
```


CHANGELOG.md
## 0.10.0

- Simplified line strip rendering; `TransparentLineStrip` and `OpaqueLineStrip` are deprecated.

## 0.9.0

- Renamed `Grid.info` and `OrthoAxis.info` to `Grid.computed` and `OrthoAxis.computed` respectively.
- Added support for using multiple draw calls per primitive. If you've been
  using a custom primitive, you'll need to do the following:
  - Rename your `command` function to `commands`.
  - Return a `Record<string, DrawCommand>` from `commands`, where the string is
    a name you give each command. The exported type alias `NamedDrawCommands`
    resolves to this type.
  - Update your primitive's `render` function to accept a `NamedDrawCommands`
    and use the names you provided to find the `DrawCommand` you need.
- Added the `TransparentLineStrip` primitive that allows rendering line strips
  with alpha and avoids the overlap artifacts present in `LineStrip`.
- Added the `OpaqueLineStrip` primitive and a deprecation warning when using
  `LineStrip`.

([#44](https://github.com/wwwtyro/candygraph/pull/44))

## 0.8.0

- Refactored all resources from the "create" pattern to classes instantiated
  with `new`. E.g., `createLineStrip(...)` is now `new LineStrip(...)`.
  ([#42](https://github.com/wwwtyro/candygraph/pull/42))

## 0.7.0

- Fixed the blending so that compositing with the webpage looks correct. While
  technically a bugfix, we can't be sure people haven't come to rely on the
  previous blending, so this is a zero-based major version bump.
  ([#41](https://github.com/wwwtyro/candygraph/pull/41))

## 0.6.0

- Cartesian and polar coordinate systems now allow their scopes to be garbage
  collected, removing a memory leak. This requires that a CandyGraph object be
  provided upon creation. ([#37](https://github.com/wwwtyro/candygraph/pull/37))

## 0.5.1

- Fix an error when trying to create text that contains chars that are not
  encoded in the given font
  ([#31](https://github.com/wwwtyro/candygraph/pull/31))

## 0.5.0

- Removed the memory management mechanics around `retain()`. All renderables are
  now retained by default. Renderables that go out of scope will be garbage
  collected, or one may call their associated `dispose()` functions to perform
  an immediate manual cleanup.
  ([#26](https://github.com/wwwtyro/candygraph/pull/26))


src/candygraph.ts
import REGL, { DrawCommand, Regl, Vec4 } from "regl";
import { CoordinateSystem } from "./coordinates/coordinate-system";
import { Viewport, Renderable } from "./common";
import { Primitive, NamedDrawCommands } from "./primitives/primitive";
import { Composite } from "./composites/composite";

interface Props {
  resolution: [number, number];
  viewport: Viewport;
}

const commonGLSL = `
  uniform vec2 resolution;

  vec4 rangeToClip(vec2 v) {
    return vec4(2.0 * v / resolution - 1.0, 0.0, 1.0);
  }

  vec4 domainToClip(vec2 v) {
    return rangeToClip(toRange(v));
  }
`;

export interface CandyGraphOptions {
  /** The canvas element the webgl context will be created from. One will be created if not provided. */
  canvas?: HTMLCanvasElement;

  /** Indicates if the canvas contains an alpha buffer. Enable this if you would like to composite your graph with DOM elements. */
  alpha?: boolean;
}

const DEFAULT_OPTIONS = {
  canvas: null,
  alpha: false,
};

export class CandyGraph {
  /** @internal */
  public readonly regl: Regl;

  public readonly canvas: HTMLCanvasElement;

  private commandsByCoords = new Map<string, Map<Function, NamedDrawCommands>>();
  private scope: DrawCommand;

  constructor(options: CandyGraphOptions = {}) {
    const opts = { ...DEFAULT_OPTIONS, ...options };
    this.canvas = opts.canvas ?? document.createElement("canvas");
    this.regl = REGL({
      canvas: this.canvas,
      extensions: ["angle_instanced_arrays", "oes_standard_derivatives"],
      attributes: {
        depth: false,
        alpha: opts.alpha,
        premultipliedAlpha: true,
        stencil: true
      },
    });
    this.scope = this.regl({
      uniforms: {
        resolution: this.regl.prop<Props, "resolution">("resolution"),
      },
      viewport: this.regl.prop<Props, "viewport">("viewport"),
      blend: {
        enable: true,
        func: { srcRGB: "src alpha", dstRGB: "one minus src alpha", srcAlpha: "one", dstAlpha: "one minus src alpha" },
      },
      depth: {
        enable: false,
      },
    });
  }

  /** Clears the entire CandyGraph canvas.
   * @examples
   * Clear the canvas to solid white.
   * ```ts
   * cg.clear([1, 1, 1, 1]);
   * ```
   * Clear the canvas to blue.
   * ```ts
   * cg.clear([0, 0, 1, 1]);
   * ```
   * Clear the canvas to zero alpha for compositing with the page.
   * ```ts
   * const cg = new CandyGraph({ alpha: true });
   * cg.clear([0, 0, 0, 0]);
   * ```
   */
  public clear = (color: [number, number, number, number]): void => {
    this.regl.clear({ color: color as Vec4 });
  };

  /** Renders the given Renderable(s) to the given Viewport with the provided CoordinateSystem.
   * @example
   * ```ts
   * // Create a viewport.
   * const viewport = { x: 0, y: 0, width: 384, height: 384 };
   *
   * // Create a coordinate system.
   * const coords = new CartesianCoordinateSystem(
   *   new LinearScale([0, 1], [32, viewport.width - 16]),
   *   new LinearScale([0, 1], [32, viewport.height - 16])
   * );
   *
   * // Render a line segment to the viewport.
   * cg.render(coords, viewport, new LineSegments(cg, [0, 0, 1, 1]));
   *
   * // Render a couple more line segments to the viewport.
   * cg.render(coords, viewport, [new LineSegments(cg, [0.5, 0, 0.5, 1]), new LineSegments(cg, [0, 1, 1, 0])]);
   * ```
   */
  public render = (coords: CoordinateSystem, viewport: Viewport, renderable: Renderable): void => {
    coords.scope(coords.props(), () => {
      this.scope(
        {
          resolution: [viewport.width, viewport.height],
          viewport,
        },
        () => {
          this.recursiveRender(coords, renderable);
        }
      );
    });
  };

  /** Destroys the underlying `regl` instance and renders this `CandyGraph` instance unusable. */
  public destroy() {
    this.regl.destroy();
  }

  /** Copies the contents of the CandyGraph canvas to another canvas. Returns the HTMLCanvasElement that was copied to.
   * @param sourceViewport The `Viewport` of the `CandyGraph` canvas that will be copied from.
   * @param destinationCanvas The canvas that will be copied to. If not provided, one will be created with the dimensions of `destinationViewport`.
   * @param destinationViewport If not provided, one will be created that is positioned at [0, 0] and with the width and height of `sourceViewport`.
   */
  public copyTo(sourceViewport: Viewport, destinationCanvas?: HTMLCanvasElement, destinationViewport?: Viewport) {
    // If we're not provided a canvas, make one.
    const dest = destinationCanvas ?? document.createElement("canvas");

    // If we're not provided a destination viewport, make one the size of the source viewport.
    destinationViewport = destinationViewport ?? {
      x: 0,
      y: 0,
      width: sourceViewport.width,
      height: sourceViewport.height,
    };

    // If we weren't provided a destination canvas, size it according to destinationViewport.
    if (!destinationCanvas) {
      dest.width = destinationViewport.x + destinationViewport.width;
      dest.height = destinationViewport.y + destinationViewport.height;
    }

    // Copy the canvas.
    const ctx = dest.getContext("2d");
    ctx?.drawImage(
      this.canvas,
      0,
      this.canvas.height - sourceViewport.height,
      sourceViewport.width,
      sourceViewport.height,
      destinationViewport.x,
      destinationViewport.y,
      destinationViewport.width,
      destinationViewport.height
    );

    return dest;
  }

  private getNamedCommands(coords: CoordinateSystem, primitive: Primitive) {
    let commandsByPrimitive = this.commandsByCoords.get(coords.glsl);
    if (!commandsByPrimitive) {
      commandsByPrimitive = new Map<Function, NamedDrawCommands>();
      this.commandsByCoords.set(coords.glsl, commandsByPrimitive);
    }
    let namedCommands = commandsByPrimitive.get(primitive.constructor);
    if (!namedCommands) {
      namedCommands = primitive.commands(coords.glsl + commonGLSL);
      commandsByPrimitive.set(primitive.constructor, namedCommands);
    }
    return namedCommands;
  }

  private recursiveRender(coords: CoordinateSystem, renderable: Renderable) {
    if (Array.isArray(renderable)) {
      for (const element of renderable) {
        this.recursiveRender(coords, element);
      }
    } else {
      if (renderable instanceof Primitive) {
        const command = this.getNamedCommands(coords, renderable);
        renderable.render(command);
      } else if (renderable instanceof Composite) {
        if (renderable.scope) {
          renderable.scope(renderable.props(coords), () => {
            this.recursiveRender(coords, renderable.children());
          });
        } else {
          this.recursiveRender(coords, renderable.children());
        }
      }
    }
  }
}
